packages/block-editor/src/hooks/block-binding-support/index.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import withBlockBindingSupport from './with-block-binding-support';
+
+export const BLOCK_BINDINGS_ALLOWED_BLOCKS = {
+	'core/paragraph': [ 'content' ],
+	'core/heading': [ 'content' ],
+	'core/image': [ 'url', 'title', 'alt' ],
+	'core/button': [ 'url', 'text', 'linkTarget' ],
+};
+
+export function isItPossibleToBindBlock( blockName ) {
+	return blockName in BLOCK_BINDINGS_ALLOWED_BLOCKS;
+}
+
+export default function extendBlockWithBoundAttributes( settings, name ) {
+	if ( ! isItPossibleToBindBlock( name ) ) {
+		return settings;
+	}
+
+	return {
+		...settings,
+		/*
+		 * Expose relevant data through
+		 * the block context.
+		 */
+		usesContext: [
+			...new Set( [
+				...( settings.usesContext || [] ),
+				'postId',
+				'postType',
+				'queryId',
+			] ),
+		],
+		edit: withBlockBindingSupport( settings.edit ),
+	};
+}
+
+addFilter(
+	'blocks.registerBlockType',
```

Comment:
As part of the refactoring, I wanted to review if it is possible to use the syntax other hooks are using: [link](https://github.com/WordPress/gutenberg/blob/a988b90e855211f9094a1c2abe3218bb6ea8d61c/packages/block-editor/src/hooks/align.js#L154-L163). They just export an object instead of using `addFilter` and `blocks.registerBlockType`. I believe that improves the performance of the hooks.

I believe with this refactoring, it should be feasible.

Reply:
It would be important to follow how other filters are wired so all the performance improvements added by @ellatrix don't get obsolete over time.

Reply:
Right, we should closely monitor the performance of vitals

I know we're avoiding adding more logic/complexity to the [`BlockListBlock`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/block.js), but maybe that's the right place for this API. I would love to hear what others think about this.

cc @youknowriad

Reply:
Could we address this ina follow-up?

Reply:
> As part of the refactoring, I wanted to review if it is possible to use the syntax other hooks are using: [link](https://github.com/WordPress/gutenberg/blob/a988b90e855211f9094a1c2abe3218bb6ea8d61c/packages/block-editor/src/hooks/align.js#L154-L163). They just export an object instead of using addFilter and blocks.registerBlockType.

TIL about this format! 😃  We should address this in the current PR. 

> I know we're avoiding adding more logic/complexity to the [BlockListBlock](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/block.js), but maybe that's the right place for this API. I would love to hear what others think about this.

This part I would say is totally fine for a follow-up but I'm still curious to hear from Riad to hear about how it might be done 🙂

Reply:
> > As part of the refactoring, I wanted to review if it is possible to use the syntax other hooks are using: [link](https://github.com/WordPress/gutenberg/blob/a988b90e855211f9094a1c2abe3218bb6ea8d61c/packages/block-editor/src/hooks/align.js#L154-L163). They just export an object instead of using addFilter and blocks.registerBlockType.
> 
> TIL about this format! 😃 We should address this in the current PR.

The properties passed down to the `edit` don't contains the `attributes` one, required to bind with the external source. I was tempted to add it, but [then](https://github.com/WordPress/gutenberg/blob/56dc165707ebc832967e9026b090a2690d4399f9/packages/block-editor/src/hooks/utils.js#L468-L471)...

```jsx
// This component is pure, so only pass needed
// props!!!
{ ...neededProps }
```

I'm not sure if we could pass the `attributes` prop there.

Reply:
After researching a little I'm not sure if we can use `createBlockEditFilter`. We need to define explicitely the attributes that we want to expose when extending the edit function. We know ahead we need to add the `metadata` attribute there:


```
export default {
	edit: BlockBindingBridge,
	attributeKeys: [ 'metadata' ],
	hasSupport: isItPossibleToBindBlock,
};
```

But also we need to add the block attributes bound with the external source, which are defined as part of the metadata attribute ♾️

Reply:
Michal, this is [an experimental](https://github.com/WordPress/gutenberg/pull/58895/commits/853483c9e96e4c61eb8b019be8f3e43ad8cbb38b) approach to use the hooks API to extend the block with bound attrs:

* It handles the especial "metadata" attribute case, exposing not only the `metadata` attribute defined by the `attributeKeys`, but also all attr dependencies. Meaning that if the block binds the `content` attribute, it will be picked and exposed in the block props by the Hook API.
* To make the implementation simpler (debatable 😅 ), it also maps the attribute value. Meaning that in the block edit context, the `metadata` block property will have the following shape:

```json
{
  "bindings": {
    "<attr-name>": {
      "source": "<source-handler>",
      "key": "<source-prop>",
      "value": "<attr-value>"
    },
    // ...
  }
}
```

In our classic example:

```json
{
  "bindings": {
    "content": {
      "source": "core/post-meta",
      "key": "text_custom_field",
      "value": "Hello, Bound world"
    },
  }
}
```
* it also propagates the context by probably it should be picked from the React context (todo)